### DISCUSSION TITLE
Can I modify the parameters  after loading udrf？

### DISCUSSION THREAD

**Discussion by lcb512** [stack-of-tasks/pinocchio#1910] [Q&A] [answered]
I want to do experiments on exoskeleton robots with different leg lengths. Can I modify the parameters (size, quality) after loading udrf？

**Comment by jcarpent** [comment]
Yes, of course, you can modify your model after loading: change inertia, joint placements, etc.

**Comment by lcb512** [comment, marked as answer]
I want to modify the leg length in rqt instead of directly modifying it in urdf. I don't know what to do. Is there any good way?

**Reply by jcarpent** [reply]
I don't know what is rqt.

**Reply by lcb512** [reply]
software interface.

**Reply by jcarpent** [reply]
You should provide the interface and rqt. There is no direct bridge.

**Reply by stephane-caron** [reply]
@lcb512 You should give more details to get helpful answers. For example here: [rqt](https://wiki.ros.org/rqt) is a Qt-based framework for GUI development for ROS. It enables the development of custom interfaces that look like this:

![image](https://user-images.githubusercontent.com/1189580/229444749-729a24b1-b7aa-41c4-905f-11b91b308396.png)

In your case whether you use rqt or any other form of user input is not relevant. If you want to modify the leg length, what's relevant in @jcarpent's answer is that you can update **joint placements** of your robot after loading it from URDF.

**Reply by jcarpent** [reply]
Following the nice answer of @stephane-caron, in Python you can easily do:
```python
model.jointPlacements[model.getJointId("my_joint_name")].translation = np.array([1.,2.,3.])
```

**Reply by lcb512** [reply]
Ok, I understand,what is the code for C++.

**Reply by jcarpent** [reply]
Exactly the same, except you should use Eigen:
```cpp
model.jointPlacements[model.getJointId("my_joint_name")].translation << 1.,2.,3.;
```

**Reply by lcb512** [reply]
Thank you very much. I'll try it.

**Reply by stephane-caron** [reply]
Proposing an example to illustrate this use case, you can check out the code (Python) in this PR: https://github.com/stack-of-tasks/pinocchio/pull/1911/files

```python
import numpy as np
from robot_descriptions.loaders.pinocchio import load_robot_description  # pip install robot_descriptions

robot = load_robot_description("upkie_description")
model = robot.model

# Joint placement offsets compared to the leg length
# We know from the robot description that both are along the local y-axis
known_offsets = {"knee": 0.072, "wheel": 0.065}

new_length = 0.3  # in meters, length for both femur and tibia links
for side in ("left", "right"):
    for joint in ("knee", "wheel"):
        joint_id = model.getJointId(f"{side}_{joint}")
        model.jointPlacements[joint_id].translation[1] = new_length - known_offsets[joint]
```
